src/AvoDeduplicator.ts

```diff
@@ -0,0 +1,183 @@
+import { AvoSchemaParser } from "./AvoSchemaParser";
+import { deepEquals } from "./utils";
+
+export class AvoDeduplicator {
+  avoFunctionsEvents: { [time: number]: string } = {};
+  manualEvents: { [time: number]: string } = {};
+
+  avoFunctionsEventsParams: {
+    [eventName: string]: { [propName: string]: any };
+  } = {};
+  manualEventsParams: { [eventName: string]: { [propName: string]: any } } = {};
+
+  shouldRegisterEvent(
+    eventName: string,
+    params: { [propName: string]: any },
+    fromAvoFunction: boolean
+  ): boolean {
+    this.clearOldEvents();
+
+    if (fromAvoFunction) {
+      this.avoFunctionsEvents[Date.now()] = eventName;
+      this.avoFunctionsEventsParams[eventName] = params;
+    } else {
+      this.manualEvents[Date.now()] = eventName;
+      this.manualEventsParams[eventName] = params;
+    }
+
+    let checkInAvoFunctions = !fromAvoFunction;
+
+    return !this.hasSameEventAs(eventName, params, checkInAvoFunctions);
+  }
+
+  private hasSameEventAs(
+    eventName: string,
+    params: { [propName: string]: any },
+    checkInAvoFunctions: boolean
+  ): boolean {
+    let result = false;
+
+    if (checkInAvoFunctions) {
+      for (const otherEventName in this.avoFunctionsEventsParams) {
+        if (this.avoFunctionsEventsParams.hasOwnProperty(otherEventName)) {
+          if (otherEventName === eventName) {
+            const otherParams = this.avoFunctionsEventsParams[eventName];
+            if (deepEquals(params, otherParams)) {
+              result = true;
+              break;
+            }
+          }
+        }
+      }
+    } else {
+      for (const otherEventName in this.manualEventsParams) {
+        if (this.manualEventsParams.hasOwnProperty(otherEventName)) {
+          if (otherEventName === eventName) {
+            const otherParams = this.manualEventsParams[eventName];
+            if (deepEquals(params, otherParams)) {
+              result = true;
+              break;
+            }
+          }
+        }
+      }
```

Comment:
Maybe extract to a function to reuse the logic between the two?

Reply:
maybe...